widget/table.go

```diff
@@ -0,0 +1,299 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+)
+
+// Table widget is a grid of items that can be scrolled and a cell selected.
+// It's performance is provided by caching cell templates created with NewCell and re-using them with UpdateCell.
+// The size of the content rows/columns is returned by the DataSize callback.
+type Table struct {
+	BaseWidget
+
+	DataSize       func() (int, int)
```

Comment:
It maybe longer, but `Dimensions` might be a better name

Reply:
That is a good point. But would that imply that List should have been Dimension?

I was thinking about the semantics and it's tough, does Dimensions (plural of dimension) refer to many different dimensions, whereas this is sort of 1 dimension that is two dimensional ;).

Reply:
Hmmm, interesting. To flip it around, if this is DataSize, then should list also be DataSize?

I think it would be okay to have List.Length and Table.Dimensions, but I also like the consistency of List.Dimension and Table.Dimensions, @okratitan wdyt?

Reply:
I think Dimension implies spatial undertones and for list that is irrelevant.  When I hear Dimension I think "Okay so with dimension I must be defining if this list should be vertical or horizontal" more so than I think about it as "How many" which it is  For table using dimensions makes sense because you are defining "10 columns" by "10 rows" which is spatial and not the how many... "How many" is 100.  I do think dimensions sounds better than data size for list.

Reply:
I don't know if Dimension is clearly separated from the concept of widget size enough - it could be deemed ambiguous.
Length for List works in this way, but can we name this in a way that does not suffer that potential confusion?

Reply:
I do agree that it makes sense for both list and table to use similar terminology.

I think that Size() makes more sense than dimension, and would also apply to list as well. 

Could also have two separate calls, a `NumberOfRows()`, and `NumberOfCols()` (or just Rows and Cols). I think `Rows()` also makes sense in the context of list. That said, the short versions (Rows and Cols) may be confused with a method to actually retrieve the data of the rows/cols, rather than the number of rows/cols.